design-proposals/klip-23-primary-keys-for-tables.md

```diff
@@ -0,0 +1,161 @@
+* Only syntax for a single PRIMARY KEY.
+
+## What is not in scope
+
+* Any functional change in key handling semantics. The proposed change is purely syntactical.
+* Multiple key columns. Syntax for composite primary keys is well documented in the SQL community
+and will be introduced with the work to support multiple key columns.
+
+## Value/Return
+
+Main gains:
+
+1. More standard SQL: `PRIMARY KEY` is standard sql syntax, where as `KEY` is not.
+2. Should help differentiate the difference between a stream's key column and a table's primary key
+to users.
+
+There is also some discussion going on about how ksqlDB should model streams: should they continue
+to be their own collection type, or should they be modelled as tables without primary keys? If we
+do go the latter route it will be crucial that we differentiate `PRIMARY KEY`s from non-primary key
+columns.
+
+## Public APIS
+
+`CREATE TABLE` statements will change from using the `KEY` keyword for key columns to `PRIMARY KEY`.
+For example:
+
+```sql
+-- current syntax:
+CREATE TABLE FOO (ID INT KEY, NAME STRING) ...
+
+-- proposed syntax:
+CREATE TABLE FOO (ID INT PRIMARY KEY, NAME STRING) ...
+```
+
+Note: `CREATE STREAM` statements will not be affected.
+
+## Design
+
+This is purely a syntax change.  All that is needed is:
+
+ 1. add an optional `PRIMARY` keyword to be added to the syntax of create statements, and
+ 1. to have the parser reject any `CREATE TABLE` key columns _without_ the `PRIMARY`
+keyword, and reject any `CREATE STREAM` key column _with_ the new keyword.
+
+## Test plan
+
+QTT tests will be updated to reflect the new syntax for tables. Existing historic tests will ensure
+the change is backwards compatible.
+
+## LOEs and Delivery Milestones
+
+Change already implemented: https://github.com/confluentinc/ksql/pull/4986.
+
+## Documentation Updates
+
+See existing PR for doc updates: https://github.com/confluentinc/ksql/pull/4986
+
+This new syntax will be highlighted in a blog post, (tracked by
+https://github.com/confluentinc/ksql/issues/4960). The existing examples in the ksqlDB repo and the
+examples repo will need updating with the new syntax,
+(tracked by https://github.com/confluentinc/ksql/issues/4927).
+
+Where out of data examples are used in the new version of ksqlDB, the error message will be very explicit about what needs to be changed, e.g. 
+
+```
+ksql> CREATE TABLE FOO (ROWKEY INT KEY, NAME STRING) WITH (...);
+Error: Line: 1, COL: 18: `KEY` used in table schema. Tables have primary keys. Please replace KEY with PRIMARY KEY in: ROWKEY INT KEY
+
+The following columns are defined as KEY columns. Tables do not support KEY columns, only PRIMARY KEY columns. Please add the `PRIMARY` key word. Columns:
+ROWKEY
+Tables have PRIMARY KEYs, which are unique and NON NULL.
+Streams have KEYs, which have no uniqueness or NON NULL constraints.
+```
+
+## Compatibility Implications
+
+None
```

Comment:
Well, isn't the change backward incompatible if a `CTAS` using `KEY` instead of `PRIMARY KEY` is rejected?

Reply:
Meh, I've spelled it out.